wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/environment/InternalEnvironmentFactory.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server.spi.environment;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.model.workspace.Warning;
+import org.eclipse.che.api.core.model.workspace.config.Environment;
+import org.eclipse.che.api.core.model.workspace.config.MachineConfig;
+import org.eclipse.che.api.core.model.workspace.config.ServerConfig;
+import org.eclipse.che.api.installer.server.InstallerRegistry;
+import org.eclipse.che.api.installer.server.exception.InstallerException;
+import org.eclipse.che.api.installer.shared.model.Installer;
+import org.eclipse.che.api.workspace.server.model.impl.ServerConfigImpl;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+
+/**
+ * Creates a valid instance of InternalEnvironment.
+ *
+ * <p>Expected to be bound with a MapBinder with unique String key that contains recipe type, like:
+ *
+ * <pre>
+ *   MapBinder<String, InternalEnvironmentFactory> envFactories =
+ *       MapBinder.newMapBinder(binder(), String.class, InternalEnvironmentFactory.class);
+ *   envFactories.addBinding("recipe_type_1").to(SubclassOfInternalEnvironmentFactory.class);
+ * </pre>
+ *
+ * @author gazarenkov
+ * @author Sergii Leshchenko
+ */
+public abstract class InternalEnvironmentFactory<T extends InternalEnvironment> {
+
+  private final InstallerRegistry installerRegistry;
+  private final RecipeRetriever recipeRetriever;
+  private final InternalEnvironmentValidator<T> envValidator;
+
+  public InternalEnvironmentFactory(
+      InstallerRegistry installerRegistry,
+      RecipeRetriever recipeRetriever,
+      InternalEnvironmentValidator<T> envValidator) {
+    this.installerRegistry = installerRegistry;
+    this.recipeRetriever = recipeRetriever;
+    this.envValidator = envValidator;
+  }
+
+  /**
+   * Creates a valid instance of InternalEnvironment.
```

Comment:
Can you add more details on what is performed inside of this method? Looks like it is much more than just an instance creation

Reply:
Added describing of logic that is performed while instance creation. Please take a look if it is clear enough.